src/nemos/basis/_basis_mixin.py

```diff
@@ -255,4 +489,182 @@ def to_transformer(self) -> TransformerBasis:
+        # in the composite basis.
+        self.basis1 = copy.deepcopy(basis1)
+        self.basis2 = copy.deepcopy(basis2)
+
+        # set parents
+        self.basis1._parent = self
+        self.basis2._parent = self
+
+        shapes = (
+            *(bas1._input_shape_ for bas1 in basis1._iterate_over_components()),
+            *(bas2._input_shape_ for bas2 in basis2._iterate_over_components()),
+        )
+        # if all bases where set, then set input for composition.
+        set_bases = [s is not None for s in shapes]
+
+        if all(set_bases):
+            # pass down the input shapes
+            self.set_input_shape(*shapes)
+        elif any(set_bases):
+            warnings.warn(
+                "Only some of the basis where initialized with `set_input_shape`, "
+                "please initialize the composite basis before computing features.",
+                category=UserWarning,
+            )
+
+    @property
+    @abc.abstractmethod
+    def n_basis_funcs(self):
+        """Read only property for composite bases."""
+        pass
+
+    def setup_basis(self, *xi: NDArray) -> Basis:
+        """
+        Set all basis states.
+
+        This method corresponds sklearn transformer ``fit``. As fit, it must receive the input and
+        it must set all basis states, i.e. ``kernel_`` and all the states relative to the input shape.
+        The difference between this method and the transformer ``fit`` is in the expected input structure,
+        where the transformer ``fit`` method requires the inputs to be concatenated in a 2D array, while here
+        each input is provided as a separate time series for each basis element.
+
+        Parameters
+        ----------
+        xi:
+            Input arrays.
+
+        Returns
+        -------
+        :
+            The basis with ready for evaluation.
+        """
+        # setup both input independent
+        self._set_input_independent_states()
+
+        # and input dependent states
+        self.set_input_shape(*xi)
+
+        return self
+
+    def _set_input_independent_states(self):
+        """
+        Compute the input dependent states for traversing the composite basis.
+
+        Returns
+        -------
+        :
+            The basis with the states stored as attributes of each component.
+        """
+        self.basis1._set_input_independent_states()
+        self.basis2._set_input_independent_states()
+
+    def _check_input_shape_consistency(self, *xi: NDArray):
+        """Check the input shape consistency for all basis elements."""
+        self._basis1._check_input_shape_consistency(
+            *xi[: self._basis1._n_input_dimensionality]
+        )
+        self._basis2._check_input_shape_consistency(
+            *xi[self._basis1._n_input_dimensionality :]
+        )
+
+    @property
+    def basis1(self):
+        return self._basis1
+
+    @basis1.setter
+    def basis1(self, bas: Basis):
+        self._basis1 = bas
```

Comment:
why are these properties if we do nothing special in the getter or setter? why not just use `self.basis1` instead of `self._basis1`?